### DISCUSSION TITLE
Help for Auxkernel of Nodal value

### DISCUSSION THREAD

**Discussion by bylore** [idaholab/moose#26553] [Q&A General] [answered]
Hi, I am trying to implemente Configurational force in Moose.
Configuration force in finite element method is a nodal value which can be calculated by the following formulation. 
![image](https://github.com/idaholab/moose/assets/18157950/fb3ee96a-3693-46da-ae60-59a8d02c9a7b)
It is a integral over the element e, and N^I is the shape function correspnding to node I.
G_e^I is the configuration force contributing to node I from element e.
![image](https://github.com/idaholab/moose/assets/18157950/56d8f05d-835e-45b5-a8b2-7906067a279d)

I think Auxkernel can compute some nodal values, but i don't know how to add  the nodal values contributing from different adjacent elements together. Could you give some advice? Thank you!

Ref:[1] R. Mueller, G. A. Maugin, Comput. Mech. 2002, 29, 52.

**Comment by GiudGiud** [comment, marked as answer]
Hello

I think you should be able to use the code developed for nodal patch recovery
https://mooseframework.inl.gov/help/finite_element_concepts/nodal_patch_recovery.html
to do this.

look at the following source files for how to do this
[framework/src/auxkernels/NodalPatchRecovery.C](https://github.com/idaholab/moose/blob/next/framework/src/auxkernels/NodalPatchRecovery.C)
[framework/src/userobjects/NodalPatchRecoveryBase.C](https://github.com/idaholab/moose/blob/next/framework/src/userobjects/NodalPatchRecoveryBase.C)

Guillaume

**Reply by bylore** [reply]
Great，Thank you！

**Reply by bylore** [reply]
Hello
In the source code of NodalPatchRecovery.C, it says that

```
// Limit current use of NodalPatchRecovery to a single processor
  if (_communicator.size() > 1)
    mooseError("The nodal patch recovery option, which calculates the Zienkiewicz-Zhu patch "
               "recovery for nodal variables (family = LAGRANGE), is not currently implemented for "
               "parallel runs. Run in serial if you must use the nodal patch capability");
```
If I only utilize some of its code to compute the node value from it element patch as above. Should i worried about this or Can i calculate this for parallel run?

**Reply by bylore** [reply]
NodalPatchRecoveryAux.C seems dosen't have such limitations.

**Reply by GiudGiud** [reply]
@hugary1995

**Reply by hugary1995** [reply]
Nodal patch recovery (NPR) and the configurational force (CF) are different. NPR reconstructs a multi-indexing polynomial basis which is then used to interpolate the value at each node. CF assembles all adjacent element integrals into a set.

What they have in common is that they both need to loop over all the adjacent elements of each node. So I would second @GiudGiud 's suggestion -- those two source files are a good place to start.

**Reply by bylore** [reply]
Thanks, i will try it.  An alternative way to calculate nodal configuration force may be : treat it as a Kernel of one arbitrary variable and save this residual in nodal auxiliary variable and cancel the residual of this kernel to the chosen variable by adding a corresponding negative one.

**Reply by GiudGiud** [reply]
AuxKernel storing into an auxvariable will let you compute the contribution to the residual, but not the contribution to the Jacobian (no AD in auxiliary variables to start with).

You could look at doing that with a material property though
